### PR TITLE
Sync blockchain before running e2e tests on public testnet #568

### LINKED ISSUE #568: Sync blockchain before running e2e tests on public testnet
# Problem

If e2e tests start on a public testnet and the blockchain is not synced, the tests fail due to timeout.

# Implementation

Have a prerequisite that checks if the blockchain is synced. If it's not, sync it before running e2e tests.

Instead of:
```
make test-e2e networkid=4
```

Do something like:
```
./build/bin/statusd -les -networkid 4 -exit-when-synced
make test-e2e networkid=4
```

This needs to be fixed in `.travis.yml`.

# Acceptance Criteria

1. E2e tests do not fail because of not synced blockchain.

### PR DESCRIPTION
**Pull request by canercidam**
Option to sync blockchain and exit in `statusd`, for synchronizing before e2e tests.

Changes:
- `-exit-when-synced` flag is added to `statusd`.
- `NodeManager` is extended with `Sync`.
- `Sync` reuses the code in "testing/testing.go", `EnsureNodeSync`.
- `.travis.yml` is modified to sync before e2e tests on public network.

Closes #568

### CODE REVIEW

File: cmd/statusd/main.go [hunk 1]
```diff
@@ -62,6 +62,8 @@ var (
 	// Push Notification
 	enablePN     = flag.Bool("shh.notify", false, "Node is capable of sending Push Notifications")
 	firebaseAuth = flag.String("shh.firebaseauth", "", "FCM Authorization Key used for sending Push Notifications")
+
+	exitWhenSynced = flag.Bool("exit-when-synced", false, "Exit after blockchain sync")
```
Comment (adambabik) [inline cmd/statusd/main.go]: Maybe we can make it an integer and default to `-1`? If it's larger than `-1` then this option is active and its value tells how long to wait for a synchronization to finish? `0` could mean wait infinitely long until it's synced.

File: cmd/statusd/main.go [hunk 2]
```diff
@@ -106,6 +108,12 @@ func main() {
 		go startCollectingStats(interruptCh, backend.NodeManager())
 	}
 
+	// Sync blockchain and stop.
+	if *exitWhenSynced {
+		syncStopErr := backend.NodeManager().SyncAndStopNode(time.Minute * 50)
+		defer handleSyncStopErr(syncStopErr)
```
Comment (adambabik) [inline cmd/statusd/main.go]: I am not really sure why `defer` was chosen. I would suggest making `SyncAndStopNode` synchronous and actually, change this method a bit (more about that below).

```go
err := backend.NodeManager().SyncAndStopNode(*exitWhenSynced * time.Minute)
if err != nil {
  log.Fatalf("Failed to sync: %v", err)
}
return
```

File: geth/node/manager.go [hunk 3]
```diff
@@ -518,3 +519,77 @@ func (m *NodeManager) isNodeAvailable() error {
 
 	return nil
 }
+
+// SyncAndStopNode waits until node synchronization and exits.
+// Reports errors to returned channel.
+func (m *NodeManager) SyncAndStopNode(timeout time.Duration) (errChan <-chan error) {
```
Comment (adambabik) [inline geth/node/manager.go]: I don't see any benefits in having a single method that does all this stuff. There is a method to stop a node so I guess only `Sync()` would be needed. In `main.go`, it can be called like this:
```go
if err := m.Sync(); err != nil {
  // exit
}
if err := m.StopNode(); err != nil {
  // exit
}
```

Comment (adambabik) [reply geth/node/manager.go]: Also, it should have a synchronous interface. If the caller wants to call it asynchronously, then it's its trouble.

File: geth/node/manager.go [hunk 4]
```diff
@@ -518,3 +519,77 @@ func (m *NodeManager) isNodeAvailable() error {
...
+		return
+	}
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-timer.C:
```
Comment (adambabik) [inline geth/node/manager.go]: I would use `<-time.After(timeout)` as it's used just once in contrast to `ticker`.

**Review by adambabik — CHANGES_REQUESTED** [review]
Thanks for the PR! Overall, it's the way to go but there are a few things that could be improved.

File: cmd/statusd/main.go (hunk 1 above)
Comment (canercidam) [reply cmd/statusd/main.go]: Sure, shall we expect minutes for values larger than 0?

Comment (canercidam) [reply cmd/statusd/main.go]: Ah okay, it's already clear in the comment below.

File: cmd/statusd/main.go (hunk 2 above)
Comment (canercidam) [reply cmd/statusd/main.go]: I just wanted it to play along with the existing flow and make it such as:
1) Start node.
2) Launch sync asynchronously.
3) Wait until node is stopped.
4) Handle aggregated errors from sync (max 2).

So I used `defer` there to handle aggregated errors before exiting the program and it makes the least amount of modification to main function thus more readable.

File: geth/node/manager.go (hunk 4 above)
Comment (canercidam) [reply geth/node/manager.go]: :+1:

**Comment by canercidam** [comment]
@adambabik I think you're right about reducing it to `Sync` and making it synchronous. It's cleaner and also yields much less trouble. Also makes more sense in the interface as an extra function.

File: cmd/statusd/main.go (hunk 2 above)
Comment (canercidam) [reply cmd/statusd/main.go]: I think I missed the fact that reused code just tracks the progress, sorry about that. :)

File: cmd/statusd/main.go [hunk 5]
```diff
@@ -62,6 +62,8 @@ var (
 	// Push Notification
 	enablePN     = flag.Bool("shh.notify", false, "Node is capable of sending Push Notifications")
 	firebaseAuth = flag.String("shh.firebaseauth", "", "FCM Authorization Key used for sending Push Notifications")
+
+	exitWhenSynced = flag.Int("exit-when-synced", -1, "Timeout for blockchain sync and exit in minutes, zero is infinite")
```
Comment (dshulyak) [inline cmd/statusd/main.go]: name is a bit too long, could it be sync-only?

Comment (canercidam) [reply cmd/statusd/main.go]: Sure, changing.

Comment (mandrigin) [inline cmd/statusd/main.go]: That's a separate mode of an app. Maybe we should add this to the documentation.
"Special mode of the app to sync blockchain and exit when synced. Timeout is  N seconds. If N is zero, timeout is infinite"

The flag name was probably specified in the task, but I'd rather call it `-sync-and-exit`.
```bash
> statusd -sync-and-exit
```
is more readable to my eyes, IMO.

Comment (mandrigin) [reply cmd/statusd/main.go]: Also, please note that the waiting time is in minutes, that is not obvious without reading code.

File: cmd/statusd/main.go [hunk 6]
```diff
@@ -106,6 +108,21 @@ func main() {
 		go startCollectingStats(interruptCh, backend.NodeManager())
 	}
 
+	// Sync blockchain and stop.
+	if *exitWhenSynced > -1 {
```
Comment (mandrigin) [inline cmd/statusd/main.go]: I'd add a log entry there that tells that the node will just sync blockchain and then exit. I'd also probably extract this to a separate function to keep `main()` simpler.

```
if *exitWhenSynced >= 0 {
     syncAndStopNode(backend)
     return
}
```

File: geth/common/types.go [hunk 7]
```diff
@@ -53,6 +53,9 @@ type NodeManager interface {
 	// Stopped node cannot be resumed, one starts a new node instead.
 	StopNode() (<-chan struct{}, error)
 
+	// Sync waits until blockchain is synchronized.
+	Sync(timeout time.Duration) error
```
Comment (mandrigin) [inline geth/common/types.go]: It would be nice to mention that `0` means infinite timeout and not "no timeout"?
Maybe also add that it should be > 2 seconds.

File: geth/node/manager.go [hunk 8]
```diff
@@ -518,3 +519,59 @@ func (m *NodeManager) isNodeAvailable() error {
...
+	return m.sync(timeout)
+}
+
+func (m *NodeManager) sync(timeout time.Duration) error {
+	if timeout == 0 {
+		// Wait for a year (infinite).
+		timeout = time.Hour * 8765
+	}
+
+	ticker := time.NewTicker(1 * time.Second)
```
Comment (mandrigin) [inline geth/node/manager.go]: Can you please extract `1 * Second` as a constant (`tickerResolution`) and re-use it in the timeout check in `Sync`.

**Review by mandrigin — CHANGES_REQUESTED** [review]
Thanks for your submission!

In general, it looks good! It would be nice to see some small changes, mostly related to documentation.

File: cmd/statusd/main.go (hunk 5 above)
Comment (canercidam) [reply cmd/statusd/main.go]: Actually, `-sync-and-exit` looks better to me than `-sync-only`. Just because `-sync-only` is rather vague.

Comment (canercidam) [reply cmd/statusd/main.go]: About the waiting time: `statusd -h` should print the usage line which says "in minutes".  But I can also add your description to somewhere else as well if you like. Where else do you use for denoting such things in the project?

Comment (canercidam) [reply cmd/statusd/main.go]: I'll change "Timeout for blockchain sync and exit in minutes, zero is infinite" to "Timeout in minutes for blockchain sync and exit, zero is infinite" which is more intelligible.

Comment (dshulyak) [reply cmd/statusd/main.go]: any reasons not to use flag.Duration ?

File: cmd/statusd/main.go (hunk 6 above)
Comment (canercidam) [reply cmd/statusd/main.go]: Yes, sync and exit stuff look like a fairly big block of code in `main()`. I'll do both.

File: geth/common/types.go (hunk 7 above)
Comment (canercidam) [reply geth/common/types.go]: I didn't only because it's not an implementation of the function. I'll instead add it to `./geth/node/manager.go` if that works for you as well.

File: geth/node/manager.go (hunk 8 above)
Comment (canercidam) [reply geth/node/manager.go]: To be honest, I don't know why it's `1 * time.Second`. `time.Second` should already suit well. :smile:

And consider it done.

File: cmd/statusd/main.go (hunk 5 above)
Comment (canercidam) [reply cmd/statusd/main.go]: We could but I think we rely on only minutes here.

File: cmd/statusd/main.go [hunk 9]
```diff
@@ -62,6 +62,8 @@ var (
 	// Push Notification
 	enablePN     = flag.Bool("shh.notify", false, "Node is capable of sending Push Notifications")
 	firebaseAuth = flag.String("shh.firebaseauth", "", "FCM Authorization Key used for sending Push Notifications")
+
+	syncAndExit = flag.Int("sync-and-exit", -1, "Timeout in minutes for blockchain sync and exit, zero is infinite")
```
Comment (adambabik) [inline cmd/statusd/main.go]: I guess there was even a better suggestion from @mandrigin to replace `zero is infinite` with `zero means no timeout unless sync is finished`.

File: cmd/statusd/main.go [hunk 10]
```diff
@@ -106,6 +108,12 @@ func main() {
 		go startCollectingStats(interruptCh, backend.NodeManager())
 	}
 
+	// Sync blockchain and stop.
+	if *syncAndExit >= 0 {
+		syncAndStopNode(backend.NodeManager())
```
Comment (adambabik) [inline cmd/statusd/main.go]: It's better to reduce the number of functions that do `os.Exit()`. So maybe we could have `syncAndStopNode` to simply return `error` if there was any and do `log.Fatalf()` here in one place.

File: geth/node/manager.go [hunk 11]
```diff
@@ -518,3 +519,64 @@ func (m *NodeManager) isNodeAvailable() error {
...
+const tickerResolution = time.Second
+
+// Sync waits until blockchain synchronization and returns.
+// Timeout must be at least two seconds and zero timeout
+// means waiting infinitely.
+func (m *NodeManager) Sync(timeout time.Duration) error {
+	// We need to have a larger timeout than ticker delta here
+	// unless we use zero which means infinite timeout.
+	if timeout < tickerResolution*2 && timeout != 0 {
+		return errors.New("Sync timeout can only be zero (infinite) or at least two seconds")
```
Comment (adambabik) [inline geth/node/manager.go]: Usually, errors start with a lower case letter for easier joining (`log.Fatalf("sync error: %v", err)`) and as we talked in a different thread, maybe we can rephrase it to: `invalid timeout that can be zero (meaning no timeout) or greater than two seconds`.

However, I would also get rid of this `at least two seconds`. If, for instance, the chain is already synced, I should not wait at all. It can be easily achieved by running a check before `for` loop. It may require running some checks on `downloader` in a separate method as well.

File: geth/node/manager.go [hunk 12]
```diff
@@ -518,3 +519,64 @@ func (m *NodeManager) isNodeAvailable() error {
...
+	if timeout == 0 {
+		// Wait for a year (infinite).
+		timeout = time.Hour * 8765
+	}
+
+	ticker := time.NewTicker(tickerResolution)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-time.After(timeout):
```
Comment (adambabik) [inline geth/node/manager.go]: I am sorry, I remember I suggested changing it to this form but in this case, it may be better to do it differently:
```go
var timeoutCh <-chan time.Time

if timeout == 0 {
  timeoutCh = make(chan time.Time)
} else {
  timeoutCh = time.After(timeout)
}

// below
select {
case <-timeoutCh:
  // ...
case <-ticker.C:
  // ...
```

And in this way you can avoid a hackish `one year timeout` :)

**Review by adambabik — CHANGES_REQUESTED** [review]

File: cmd/statusd/main.go [hunk 13]
```diff
@@ -158,6 +166,21 @@ func startCollectingStats(interruptCh <-chan struct{}, nodeManager common.NodeMa
 	}
 }
 
+func syncAndStopNode(nodeManager common.NodeManager) {
+	log.Println("Node will synchronize and exit")
+	err := nodeManager.Sync((time.Duration)(*syncAndExit) * time.Minute)
+	if err != nil {
+		log.Fatalf("Failed while waiting for sync: %v", err)
```
Comment (mandrigin) [inline cmd/statusd/main.go]: Yeah, as @adambabik suggested, use `errors.Wrap` here and in line 179 and return error from the function.

File: geth/node/manager.go (hunk 12 above)
Comment (canercidam) [reply geth/node/manager.go]: Thank you, I was actually looking for something like this before but couldn't imagine :smile:

File: cmd/statusd/main.go (hunk 9 above)
Comment (canercidam) [reply cmd/statusd/main.go]: Sorry, changing.

File: geth/common/types.go (hunk 7 above)
Comment (mandrigin) [reply geth/common/types.go]: Yeah, but that means some terra incognita between a contract (an interface) and an implementation of the interface.
Maybe we could make a custom type to make the contract more explicit?
Currently, there are 2 cases that 1 parameter tries to match, using some implicit logic.
Can we "explicify" this logic using a custom type on top of `time.Duration`, so we can call
```go
m.Sync(common.SyncTimeout(1 * time.Minute))
m.Sync(common.SyncTimeoutInfinite)
```

Or just extract all the logic about that to the `main.go`.

One "go" way to do that would be to actually provide a context inside `Sync` function, so we can cancel it. Then we can just use `select` in `main.go` to provide the timeout functionality.

**Review by mandrigin — CHANGES_REQUESTED** [review]
Great, thanks! Just a few tiny changes.

File: geth/common/types.go (hunk 7 above)
Comment (canercidam) [reply geth/common/types.go]: @adambabik @mandrigin 
> One "go" way to do that would be to actually provide a context inside Sync function, so we can cancel it.

I think this is by far the best option here in terms of usage and contract definition. One could simply pass in `context.Background()` or `context.WithTimeout()`. That way, handling the timeout in `Sync` would be simplified and we could use the context timing logic in `main()`.

**Comment by canercidam** [comment]
@adambabik @mandrigin Would you agree renaming `Sync` to `EnsureSync` as it just tracks the progress?

File: cmd/statusd/main.go [hunk 14]
```diff
@@ -106,6 +109,13 @@ func main() {
 		go startCollectingStats(interruptCh, backend.NodeManager())
 	}
 
+	// Sync blockchain and stop.
+	if *syncAndExit >= 0 {
+		if err = syncAndStopNode(backend.NodeManager()); err != nil {
+			log.Fatalf("Node synchronization failed: %v", err)
+		}
```
Comment (mandrigin) [inline cmd/statusd/main.go]: there should be a `return`, right? :)

**Review by mandrigin — CHANGES_REQUESTED** [review]
Only 1 tiny thing left and it should be approved, I think :)

File: cmd/statusd/main.go (hunk 14 above)
Comment (adambabik) [reply cmd/statusd/main.go]: `log.Fatalf` calls `os.Exit(1)` but I would use `err :=` as there is no point in overriding previous `err` value.

File: cmd/statusd/main.go [hunk 15]
```diff
@@ -158,6 +168,28 @@ func startCollectingStats(interruptCh <-chan struct{}, nodeManager common.NodeMa
 	}
 }
 
+func syncAndStopNode(nodeManager common.NodeManager) (err error) {
+	log.Println("Node will synchronize and exit")
+	v := *syncAndExit
```
Comment (adambabik) [inline cmd/statusd/main.go]: Why `v`? :P `value` or `timeout` should be easier to reason about.

File: cmd/statusd/main.go [hunk 16]
```diff
@@ -158,6 +168,28 @@ func startCollectingStats(interruptCh <-chan struct{}, nodeManager common.NodeMa
...
+		return errors.New("negative timeout value")
+	}
+	if v == 0 {
+		err = nodeManager.EnsureSync(context.Background())
+	} else {
+		ctx, cancel := context.WithTimeout(context.Background(), (time.Duration)(v)*time.Minute)
+		err = nodeManager.EnsureSync(ctx)
+		defer cancel()
+	}
+	if err != nil {
```
Comment (adambabik) [inline cmd/statusd/main.go]: I think you can skip this error check. It's ok to properly stop a node even if it didn't manage to sync the chain.

File: cmd/statusd/main.go [hunk 17]
```diff
@@ -158,6 +168,28 @@ func startCollectingStats(interruptCh <-chan struct{}, nodeManager common.NodeMa
...
+		return
+	}
+	var done <-chan struct{}
+	done, err = nodeManager.StopNode()
```
Comment (adambabik) [inline cmd/statusd/main.go]:
```go
done, err := nodeManager.StopNode()
<-done
``` 
should work just fine.

Technically, it should also have a timeout but I think we can skip it.

**Review by adambabik — CHANGES_REQUESTED** [review]
Almost there!

File: cmd/statusd/main.go (hunk 14 above)
Comment (mandrigin) [reply cmd/statusd/main.go]: yeah, but if there is _no_ error?

File: cmd/statusd/main.go (hunk 15 above)
Comment (mandrigin) [reply cmd/statusd/main.go]: I'd even move it as a param of the function to make it stateless.

Comment (canercidam) [reply cmd/statusd/main.go]: `v` is just short for `value`. I can make the change if you like but I don't think it would make a big difference. :)

Comment (canercidam) [reply cmd/statusd/main.go]: (Basically, I just thought there's no need for a meaningful name for it in that scope.)

File: cmd/statusd/main.go (hunk 17 above)
Comment (canercidam) [reply cmd/statusd/main.go]: It returns `(err error)` so `return` is compulsory there. And I wanted it to write the error to existing var (return var).

File: cmd/statusd/main.go (hunk 16 above)
Comment (canercidam) [reply cmd/statusd/main.go]: Alright, then I definitely need to use `(github.com/pkg/errors).Wrap` here to return combined errors. Shall I `dep ensure --add` that?

Comment (mandrigin) [reply cmd/statusd/main.go]: isn't `errors.Wrap` a standard library function?

File: cmd/statusd/main.go (hunk 14 above)
Comment (canercidam) [reply cmd/statusd/main.go]: @mandrigin Thanks for pointing out. :)

I think it should return if there is no error.

Comment (canercidam) [reply cmd/statusd/main.go]: @adambabik 
> but I would use `err :=` ...

I can reflect this change in the next push.

File: cmd/statusd/main.go (hunk 16 above)
Comment (canercidam) [reply cmd/statusd/main.go]: Looks like it isn't. I also got confused for a moment: https://golang.org/pkg/errors/

File: cmd/statusd/main.go (hunk 15 above)
Comment (adambabik) [reply cmd/statusd/main.go]: I am just not a fan of one character variable names except for method receiver names. I don't see any value in shortening them but I see disadvantages :D

File: cmd/statusd/main.go (hunk 16 above)
Comment (canercidam) [reply cmd/statusd/main.go]: @adambabik @mandrigin Would you mind adding this as a dependency?

Comment (mandrigin) [reply cmd/statusd/main.go]: Yeah, I think it makes sense to add it. It is useful in many places. @adambabik what do you think?

Comment (adambabik) [reply cmd/statusd/main.go]: An alternative is moving stopping node to `main.go`. Wrapping both errors together is also not really correct because the sync error has nothing to do with stop node error.

Maybe something like this:
```go
func ensureSync(nodeManager common.NodeManager, timeout int) error {
  // it's not possible to pass negative timeout from the condition in main.go

  if timeout == 0 {
    return nodeManager.EnsureSync(context.Background())
  }

  ctx, cancel := context.WithTimeout(context.Background(), (time.Duration)(timeout)*time.Minute)
  defer cancel()

  return nodeManager.EnsureSync(ctx)
}

// in main.go
if *syncAndExit >= 0 {
  var exitCode int

  if err := ensureSync(backend.NodeManager(), *syncAndExit); err != nil {
    log.Printf("Failed to sync the chain: %v", err)
    exitCode = 1
  }

  done, err := backend.NodeManager().StopNode()
  if err != nil {
    log.Fatalf("Failed to stop node: %v", err)
  }
  <-done

  os.Exit(exitCode)
}
```

Ugh, I am not happy with this neither :(

File: cmd/statusd/main.go (hunk 15 above)
Comment (canercidam) [reply cmd/statusd/main.go]: Alright, will change this as well :)

**Review by mandrigin — APPROVED** [review]
LGTM!

**Comment by dshulyak** [comment]
@canercidam do you see any reasons why EnsureSync can't be reused for tests helper?

**Comment by canercidam** [comment]
@dshulyak Now that `NodeManager` has an implementation of it, it doesn't seem to make sense to have that helper I guess.

File: cmd/statusd/main.go [hunk 18]
```diff
@@ -158,6 +166,34 @@ func startCollectingStats(interruptCh <-chan struct{}, nodeManager common.NodeMa
...
+		err = nodeManager.EnsureSync(ctx)
+		defer cancel()
+	}
+	if err != nil {
+		log.Printf("Sync and stop error: %v\n", err)
+		exitCode = 1
+	}
+	var done <-chan struct{}
+	done, err = nodeManager.StopNode()
+	<-done
```
Comment (adambabik) [inline cmd/statusd/main.go]: `<-done` should be after checking error `if err != nil {`.

File: cmd/statusd/main.go [hunk 19]
```diff
@@ -158,6 +166,34 @@ func startCollectingStats(interruptCh <-chan struct{}, nodeManager common.NodeMa
...
+	var err error
+	if timeout == 0 {
+		err = nodeManager.EnsureSync(context.Background())
+	} else {
+		ctx, cancel := context.WithTimeout(context.Background(), (time.Duration)(timeout)*time.Minute)
+		err = nodeManager.EnsureSync(ctx)
+		defer cancel()
+	}
+	if err != nil {
+		log.Printf("Sync and stop error: %v\n", err)
```
Comment (adambabik) [inline cmd/statusd/main.go]: Wording: `Sync error: %v` (no need for a new line when using `log`).

File: cmd/statusd/main.go [hunk 20]
```diff
@@ -158,6 +166,34 @@ func startCollectingStats(interruptCh <-chan struct{}, nodeManager common.NodeMa
...
+		exitCode = 1
+	}
+	var done <-chan struct{}
+	done, err = nodeManager.StopNode()
+	<-done
+	if err != nil {
+		log.Printf("Sync and stop err: %v\n", err)
```
Comment (adambabik) [inline cmd/statusd/main.go]: Wording: `Stop node err: %v`.

**Review by adambabik — APPROVED** [review]
Nice solution with returning `exitCode` from `syncAndStopNode`! 

Minor comments but please make sure `<-done` is called after `err` check as it may be `nil` and it will blow up.

**Comment by adambabik** [comment]
@canercidam @dshulyak if the test helper is not needed we can create a follow-up issue to clean it up and fix tests after this one is merged.

**Comment by canercidam** [comment]
@dshulyak @adambabik I'm happy to submit a PR for it.

File: cmd/statusd/main.go [hunk 21]
```diff
@@ -158,6 +166,34 @@ func startCollectingStats(interruptCh <-chan struct{}, nodeManager common.NodeMa
...
+		log.Printf("Sync error: %v", err)
+		exitCode = 1
+	}
+	var done <-chan struct{}
+	done, err = nodeManager.StopNode()
+	if err != nil {
+		log.Printf("Stop node err: %v", err)
+		exitCode = 1
+	}
+	<-done
```
Comment (adambabik) [inline cmd/statusd/main.go]: Not sure I was precise enough. If `err` occurs during `StopNode()` call, we should return from the `if` statement. This still can blow up if `err` occurred as `done` may be `nil`.

```go
var done <-chan struct{}
done, err = nodeManager.StopNode()
if err != nil {
  log.Printf("Stop node err: %v", err)
  return 1
}
<-done
return
```

Comment (canercidam) [reply cmd/statusd/main.go]: Done, sorry.

**Comment by mandrigin** [comment]
@canercidam thanks for making Status better! 🥇

**Comment by canercidam** [comment]
@mandrigin A pleasure! And thanks for your time and efforts. @adambabik @mandrigin
